fpm/src/fpm_sources.f90

```diff
@@ -0,0 +1,504 @@
+            str_ends_with(lower(src_file_names(i)%s), ".h")) then
+            dir_sources(i) = parse_c_source(src_file_names(i)%s)
+        end if
+
+        ! Exclude executables unless specified otherwise
+        exclude_source(i) = (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM)
+        if (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM .and. &
+             present(with_executables)) then
+            if (with_executables) then
+
+                exclude_source(i) = .false.
+                dir_sources(i)%exe_name = basename(src_file_names(i)%s,suffix=.false.)
+
+            end if
+        end if
+
+    end do
+
+    if (.not.allocated(sources)) then
+        sources = pack(dir_sources,.not.exclude_source)
+    else
+        sources = [sources, pack(dir_sources,.not.exclude_source)]
+    end if
+
+end subroutine add_sources_from_dir
+
+
+subroutine add_executable_sources(sources,executables,is_test)
+    ! Add sources from executable directories specified in manifest
+    ! Only allow executables that are explicitly specified in manifest
+    !
+    type(srcfile_t), allocatable, intent(inout), target :: sources(:)
+    class(executable_t), intent(in), optional :: executables(:)
+    logical, intent(in) :: is_test
+
+    integer :: i, j
+
+    type(string_t), allocatable :: exe_dirs(:)
+    logical, allocatable :: exclude_source(:)
+    type(srcfile_t), allocatable :: dir_sources(:)
+
+    call get_executable_source_dirs(exe_dirs,executables)
+
+    do i=1,size(exe_dirs)
+        call add_sources_from_dir(dir_sources,exe_dirs(i)%s, &
+                     with_executables=.true.)
+    end do
+
+    allocate(exclude_source(size(dir_sources)))
+
+    do i = 1, size(dir_sources)
+
+        ! Only allow executables in 'executables' list
+        exclude_source(i) = (dir_sources(i)%unit_type == FPM_UNIT_PROGRAM)
+        
+        do j=1,size(executables)
+            if (basename(dir_sources(i)%file_name,suffix=.true.) == &
+                                         executables(j)%main) then
+                exclude_source(i) = .false.
+                dir_sources(i)%exe_name = executables(j)%name
+                dir_sources(i)%is_test = is_test
+                exit
+            end if
+        end do
+
+    end do
+
+    if (.not.allocated(sources)) then
+        sources = pack(dir_sources,.not.exclude_source)
+    else
+        sources = [sources, pack(dir_sources,.not.exclude_source)]
+    end if
+
+end subroutine add_executable_sources
+
+
+subroutine get_executable_source_dirs(exe_dirs,executables)
+    ! Build a list of unique source directories
+    !  from executables specified in manifest
+    type(string_t), allocatable, intent(inout) :: exe_dirs(:)
+    class(executable_t), intent(in) :: executables(:)
+
+    type(string_t) :: dirs_temp(size(executables))
+
+    integer :: i, j, n
+
+    n = 0
+    do i=1,size(executables)
+        if (.not.any([(dirs_temp(j)%s==executables(i)%source_dir, &
+                j=1,n)])) then
```

Comment:
This is also somewhat obscure in my opinion, maybe worth making it a function instead?